### PR TITLE
Feature/tdkn 97 serialize any property

### PR DESCRIPTION
**Pull request by bchen-talend**
**Please check if the PR fulfills these requirements**

- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
Property/Properties are not serializable.

**What is the new behavior?**
1. make Property/Properties serializable
2. change the reference in ReferenceProperties not transient, as it need on runtime actually
3. make forms in PropertiesImpl transient

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
For component which use ReferenceProperties, should not mark it as transient also, as it used on runtime

### CODE REVIEW

File: daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java [hunk 1]
```diff
@@ -14,10 +14,12 @@
 
 import org.talend.daikon.i18n.TranslatableImpl;
 
+import java.io.Serializable;
+
 /**
  * Simple implementation of {@link NamedThing}
  */
-public class SimpleNamedThing extends TranslatableImpl implements NamedThing {
+public class SimpleNamedThing extends TranslatableImpl implements NamedThing, Serializable {
```
Comment (sgandon) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: Are you sure we need to make this serializable ?

File: daikon/src/test/java/org/talend/daikon/serialize/FullExampleTestUtil.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        properties.tableProp.colListBoolean.setValue(Arrays.asList(new Boolean[] { true, false, true }));
+        properties.tableProp.colListEnum.setValue(Arrays
+                .asList(new FullExampleProperties.TableProperties.ColEnum[] { FullExampleProperties.TableProperties.ColEnum.FOO,
+                        FullExampleProperties.TableProperties.ColEnum.BAR, FullExampleProperties.TableProperties.ColEnum.FOO }));
+        properties.tableProp.colListString.setValue(Arrays.asList(new String[] { "a", "b", "c" }));
+        return properties;
+    }
+
+    static public void comparePropertiesValue(Properties expectedProperties, Properties actualProperties) {
```
Comment (sgandon) [inline daikon/src/test/java/org/talend/daikon/serialize/FullExampleTestUtil.java]: use the assert word in your method, static analyser usually rely on it.
so something like : assertPropertiesValueAreEquals ?

Comment (sgandon) [inline daikon/src/test/java/org/talend/daikon/serialize/FullExampleTestUtil.java]: Are you sure the Properties.equals is not enough ?

File: daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java (hunk 1 above)
Comment (bchen-talend) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: Make it Serializable because Property extends it. Should I make displayName and title transient?

Comment (sgandon) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: ok. Yes title and display name should be transient.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/378/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/378/)

File: daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java (hunk 1 above)
Comment (bchen-talend) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: json-io will serialize displayName and title if exist, that's why it break the unit test of `PropertiesTest.testSerializeEncryptedFailure`.
and just notice it will break PresentationItem, do we have plan to get displayName for PresentationItem from properties file or need I revert the change in this time?

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/381/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/381/)

File: daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java (hunk 1 above)
Comment (sgandon) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: I am not sure to understand! do you mean that json-io will serialize the displayName even if it is *transient* ?
Also what is the pb with PresentationItem, the build seems to pass ?

Comment (bchen-talend) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: > do you mean that json-io will serialize the displayName even if it is transient ?
no, it won't with transient, but it serialize the displayName before, and for PresentationItem, the construct method is `PresentationItem(String name, String displayName)`. 
I remember there is a problem for Button, which has displayName in generated serialize json, so I guess this change may make all the button with empty displayName on client side.

**Comment by sgandon** [comment]
waiting to merge after 0.14.0 release.

**Comment by sgandon** [comment]
@bchen-talend , could you fix the conflicts so that we merge this please ?

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/24/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/24/)

File: daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java (hunk 1 above)
Comment (francisu) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: @sgandon What about the serialization of properties for the purpose of communicating them over the wire, for the web UI? Or do we handle this through another means?

Comment (sgandon) [inline daikon/src/main/java/org/talend/daikon/SimpleNamedThing.java]: yes we have our own json-schema serializer that has an implicit knowledge of the Properties.
see https://github.com/Talend/daikon/tree/master/daikon/src/main/java/org/talend/daikon/serialize/jsonschema
